File: sdk-src/native/ios/OpenIMDriverRuntime.swift
import Foundation

let OPENIM_SHUTDOWN_ERROR_CODE = NSNumber(value: -1)
let OPENIM_SHUTDOWN_ERROR_MESSAGE = "OpenIM SDK was uninitialized"

func dispatchOpenIMMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
        block()
    } else {
        DispatchQueue.main.async(execute: block)
    }
}

final class OpenIMDriverTicket {
    let taskID: Int64
    let epoch: Int64

    init(taskID: Int64, epoch: Int64) {
        self.taskID = taskID
        self.epoch = epoch
    }
}

enum OpenIMDriverState {
    case idle
    case starting
    case active
    case stopping
}

private final class OpenIMPendingCallback {
    let ticket: OpenIMDriverTicket
    let resolve: OpenIMResolveString
    let reject: OpenIMReject
    var terminalScheduled = false

    init(ticket: OpenIMDriverTicket, resolve: @escaping OpenIMResolveString, reject: @escaping OpenIMReject) {
        self.ticket = ticket
        self.resolve = resolve
        self.reject = reject
    }
}

/// The deep lifecycle/callback module behind the generated UTS façade.
/// All native callback state changes are serialized, exactly one terminal result
/// reaches UTS, and stale callbacks/events are rejected by their session epoch.
final class OpenIMDriverRuntime {
    static let shared = OpenIMDriverRuntime()

    private let serial = DispatchQueue(label: "org.openim.uts.driver")
    private var epoch: Int64 = 0
    private var nextTaskID: Int64 = 1
    private var state: OpenIMDriverState = .idle
    private var pending: [Int64: OpenIMPendingCallback] = [:]

    private init() {}

    private func invalidateEpoch(_ nextState: OpenIMDriverState) -> (Int64, [OpenIMPendingCallback]) {
        return serial.sync {
            epoch += 1
            state = nextState
            let cancelled = Array(pending.values)
            pending.removeAll()
            return (epoch, cancelled)
        }
    }

    private func rejectCancelled(_ callbacks: [OpenIMPendingCallback]) {
        if callbacks.isEmpty { return }
        dispatchOpenIMMain {
            for callback in callbacks {
                callback.reject(OPENIM_SHUTDOWN_ERROR_CODE, OPENIM_SHUTDOWN_ERROR_MESSAGE)
            }
        }
    }

    func startSession() -> Int64 {
        let transition = invalidateEpoch(.starting)
        rejectCancelled(transition.1)
        return transition.0
    }

    func markInitialized(_ sessionEpoch: Int64, _ value: Bool) {
        serial.sync {
            if epoch == sessionEpoch && state == .starting {
                state = value ? .active : .idle
            }
        }
    }

    @discardableResult
    func shutdown() -> Int64 {
        let transition = invalidateEpoch(.stopping)
        rejectCancelled(transition.1)
        return transition.0
    }

    func finishShutdown(_ stoppingEpoch: Int64) {
        serial.sync {
            if epoch == stoppingEpoch && state == .stopping {
                state = .idle
            }
        }
    }

    func register(resolve: @escaping OpenIMResolveString, reject: @escaping OpenIMReject) -> OpenIMDriverTicket {
        return serial.sync {
            let ticket = OpenIMDriverTicket(taskID: nextTaskID, epoch: epoch)
            nextTaskID += 1
            pending[ticket.taskID] = OpenIMPendingCallback(ticket: ticket, resolve: resolve, reject: reject)
            return ticket
        }
    }

    private func consumeTerminal(_ ticket: OpenIMDriverTicket) -> OpenIMPendingCallback? {
        return serial.sync {
            guard let callback = pending[ticket.taskID],
                  callback.ticket.epoch == ticket.epoch,
                  epoch == ticket.epoch,
                  callback.terminalScheduled else { return nil }
            pending.removeValue(forKey: ticket.taskID)
            return callback
        }
    }

    private func scheduleTerminal(_ ticket: OpenIMDriverTicket, _ delivery: @escaping (OpenIMPendingCallback) -> Void) {
        serial.async {
            guard let callback = self.pending[ticket.taskID],
                  callback.ticket.epoch == ticket.epoch,
                  self.epoch == ticket.epoch,
                  !callback.terminalScheduled else { return }
            callback.terminalScheduled = true
            dispatchOpenIMMain {
                guard let winner = self.consumeTerminal(ticket) else { return }
                delivery(winner)
            }
        }
    }

    func resolve(_ ticket: OpenIMDriverTicket, _ data: String) {
        scheduleTerminal(ticket) { callback in callback.resolve(data) }
    }

    func resolvePrepared(_ ticket: OpenIMDriverTicket, _ prepare: @escaping () throws -> String) {
        scheduleTerminal(ticket) { callback in
            do {
                let data = try prepare()
                callback.resolve(data)
            } catch {
                callback.reject(NSNumber(value: -1), error.localizedDescription)
            }
        }
    }

    func reject(_ ticket: OpenIMDriverTicket, _ errCode: NSNumber, _ errMsg: String) {
        scheduleTerminal(ticket) { callback in callback.reject(errCode, errMsg) }
    }

    func progress(_ ticket: OpenIMDriverTicket, _ delivery: @escaping () -> Void) {
        serial.async {
            guard let callback = self.pending[ticket.taskID],
                  callback.ticket.epoch == ticket.epoch,
                  self.epoch == ticket.epoch,
                  !callback.terminalScheduled else { return }
            dispatchOpenIMMain {
                let active = self.serial.sync { self.epoch == ticket.epoch && self.state == .active }
                if active { delivery() }
            }
        }
    }

    private func canEmitEvent(_ sessionEpoch: Int64, _ allowWhileStarting: Bool) -> Bool {
        return epoch == sessionEpoch && (state == .active || (allowWhileStarting && state == .starting))
    }

    func emitEvent(_ sessionEpoch: Int64, _ allowWhileStarting: Bool = false, _ delivery: @escaping () -> Void) {
        serial.async {
            guard self.canEmitEvent(sessionEpoch, allowWhileStarting) else { return }
            dispatchOpenIMMain {
                let active = self.serial.sync { self.canEmitEvent(sessionEpoch, allowWhileStarting) }
                if active { delivery() }
            }
        }
    }

    func pendingCountForTests() -> Int { return serial.sync { pending.count } }
    func epochForTests() -> Int64 { return serial.sync { epoch } }
    func stateForTests() -> OpenIMDriverState { return serial.sync { state } }
}

File: sdk-src/native/android/OpenIMDriverRuntime.kt
package uts.sdk.modules.unixOpenimSdk

import android.os.Handler
import android.os.Looper
import java.util.LinkedHashMap
import java.util.concurrent.Callable
import java.util.concurrent.Executors

internal const val OPENIM_SHUTDOWN_ERROR_CODE = -1
internal const val OPENIM_SHUTDOWN_ERROR_MESSAGE = "OpenIM SDK was uninitialized"

internal val openIMMainHandler = Handler(Looper.getMainLooper())

internal fun dispatchOpenIMMain(block: () -> Unit) {
  if (Looper.myLooper() == Looper.getMainLooper()) {
    block()
  } else {
    openIMMainHandler.post { block() }
  }
}

internal data class OpenIMDriverTicket(
  val taskID: Long,
  val epoch: Long
)

internal enum class OpenIMDriverState {
  IDLE,
  STARTING,
  ACTIVE,
  STOPPING
}

private class OpenIMPendingCallback(
  val ticket: OpenIMDriverTicket,
  val resolve: OpenIMResolveString,
  val reject: OpenIMReject,
  var terminalScheduled: Boolean = false
)

/**
 * The deep lifecycle/callback module behind the generated UTS façade.
 *
 * Native callbacks may arrive synchronously, on arbitrary threads, more than once,
 * or after Core teardown. All state changes pass through one serial executor. Only
 * the winning terminal callback is delivered, and delivery is always on main.
 */
internal object OpenIMDriverRuntime {
  private val serial = Executors.newSingleThreadExecutor { runnable ->
    Thread(runnable, "openim-uts-driver").apply { isDaemon = true }
  }
  private var epoch: Long = 0
  private var nextTaskID: Long = 1
  private var state: OpenIMDriverState = OpenIMDriverState.IDLE
  private val pending = LinkedHashMap<Long, OpenIMPendingCallback>()

  private fun <T> readSerial(block: () -> T): T {
    return serial.submit(Callable<T> { block() }).get()
  }

  private fun invalidateEpoch(nextState: OpenIMDriverState): Pair<Long, List<OpenIMPendingCallback>> {
    return readSerial {
      epoch += 1
      state = nextState
      val cancelled = pending.values.toList()
      pending.clear()
      Pair(epoch, cancelled)
    }
  }

  private fun rejectCancelled(callbacks: List<OpenIMPendingCallback>) {
    if (callbacks.isEmpty()) return
    dispatchOpenIMMain {
      for (callback in callbacks) {
        callback.reject(OPENIM_SHUTDOWN_ERROR_CODE, OPENIM_SHUTDOWN_ERROR_MESSAGE)
      }
    }
  }

  fun startSession(): Long {
    val transition = invalidateEpoch(OpenIMDriverState.STARTING)
    rejectCancelled(transition.second)
    return transition.first
  }

  fun markInitialized(sessionEpoch: Long, value: Boolean) {
    readSerial {
      if (epoch == sessionEpoch && state == OpenIMDriverState.STARTING) {
        state = if (value) OpenIMDriverState.ACTIVE else OpenIMDriverState.IDLE
      }
    }
  }

  fun shutdown(): Long {
    val transition = invalidateEpoch(OpenIMDriverState.STOPPING)
    rejectCancelled(transition.second)
    return transition.first
  }

  fun finishShutdown(stoppingEpoch: Long) {
    readSerial {
      if (epoch == stoppingEpoch && state == OpenIMDriverState.STOPPING) {
        state = OpenIMDriverState.IDLE
      }
    }
  }

  fun register(resolve: OpenIMResolveString, reject: OpenIMReject): OpenIMDriverTicket {
    return readSerial {
      val ticket = OpenIMDriverTicket(nextTaskID, epoch)
      nextTaskID += 1
      pending[ticket.taskID] = OpenIMPendingCallback(ticket, resolve, reject)
      ticket
    }
  }

  private fun consumeTerminal(ticket: OpenIMDriverTicket): OpenIMPendingCallback? {
    return readSerial {
      val callback = pending[ticket.taskID]
      if (callback == null || callback.ticket.epoch != ticket.epoch || epoch != ticket.epoch || !callback.terminalScheduled) {
        null
      } else {
        pending.remove(ticket.taskID)
      }
    }
  }

  private fun scheduleTerminal(ticket: OpenIMDriverTicket, delivery: (OpenIMPendingCallback) -> Unit) {
    serial.execute {
      val callback = pending[ticket.taskID]
      if (callback == null || callback.ticket.epoch != ticket.epoch || epoch != ticket.epoch || callback.terminalScheduled) {
        return@execute
      }
      callback.terminalScheduled = true
      dispatchOpenIMMain {
        val winner = consumeTerminal(ticket)
        if (winner != null) delivery(winner)
      }
    }
  }

  fun resolve(ticket: OpenIMDriverTicket, data: String) {
    scheduleTerminal(ticket) { callback -> callback.resolve(data) }
  }

  fun reject(ticket: OpenIMDriverTicket, errCode: Number, errMsg: String) {
    scheduleTerminal(ticket) { callback -> callback.reject(errCode, errMsg) }
  }

  fun progress(ticket: OpenIMDriverTicket, delivery: () -> Unit) {
    serial.execute {
      val callback = pending[ticket.taskID]
      if (callback == null || callback.ticket.epoch != ticket.epoch || epoch != ticket.epoch || callback.terminalScheduled) {
        return@execute
      }
      dispatchOpenIMMain {
        val active = readSerial { epoch == ticket.epoch && state == OpenIMDriverState.ACTIVE }
        if (active) delivery()
      }
    }
  }

  private fun canEmitEvent(sessionEpoch: Long, allowWhileStarting: Boolean): Boolean {
    return epoch == sessionEpoch && (
      state == OpenIMDriverState.ACTIVE ||
        (allowWhileStarting && state == OpenIMDriverState.STARTING)
      )
  }

  fun emitEvent(sessionEpoch: Long, allowWhileStarting: Boolean = false, delivery: () -> Unit) {
    serial.execute {
      if (!canEmitEvent(sessionEpoch, allowWhileStarting)) return@execute
      dispatchOpenIMMain {
        val active = readSerial { canEmitEvent(sessionEpoch, allowWhileStarting) }
        if (active) delivery()
      }
    }
  }

  internal fun pendingCountForTests(): Int = readSerial { pending.size }
  internal fun epochForTests(): Long = readSerial { epoch }
  internal fun stateForTests(): OpenIMDriverState = readSerial { state }
}
